### ISSUE TITLE
Problem with dev/prod imports for CSS bundles

### ISSUE THREAD

**Issue by mikhail-eremin** [systemjs/systemjs#1063]
Hi! I don't know if i should better post it to systemjs-builder repository, or to plugin-css, so i just put it here

I'm trying to use  systemjs-builder+plugin-css and this is that i see:
In plugin-css documentation, it said, i should use it like **import  './styles.css!'** (i switched to jspm to do that)
And then i do it so- build works and everything is nice! BUT only in bundled js file (inline css)
All the same doesn't work when in dev mode, i got systemjs loader error 
Error: http://sandbox.loc/styles.css!http://sandbox.loc/jspm_packages/github/systemjs/plugin-css@0.1.20.js detected as register but didn't execute.(…)

If i change  **import './styles.css!'** to **System.import("./styles.css")** it works in DEV (loads those css async), but when i'm trying to build that - i see no inlined css in my bundle

So the question is - what is the way to use css loader both in DEV and PROD?

Sorry for my English or if i asked for something dumb :)

P.S. I'm trying to build simple angularjs2 tutorial project (Heroes board)

**Comment by guybedford** [comment]
Thanks for posting a detailed explanation here.

So the first case definitely works in dev mode. If you click on the `(...)` can you expand on that error? I think you'll find it's something else, but it would be interesting to know what the catch is there.

In the second case, how did you configure the plugin? I tried it out and the bundling did load separate files in my case due to the extensions interactions. I added a fix in https://github.com/systemjs/systemjs/commit/86be5c4ac790be35ad065bef031a0f2ebc1e93ce which resolved this for me.

**Comment by mikhail-eremin** [comment]
Thanks for answering!

I already use another way for loading css in angular2 (native angular2 styles inlining), but still, loading css through "import" would be nice feature, so i configured all this again
I'm doing **import "./dashboard.component.css!";**  (path to css file http://sandbox.loc/app/dashboard/dashboard.component.css  - is 100% correct)
And this is expanded error you asked for:

Error: http://sandbox.loc/app/dashboard/dashboard.component.css!http://sandbox.loc/jspm_packages/github/systemjs/plugin-css@0.1.20.js detected as register but didn't execute.(…)run @ angular2-polyfills.js:138zoneBoundFn @ angular2-polyfills.js:111lib$es6$promise$$internal$$tryCatch @ angular2-polyfills.js:1511lib$es6$promise$$internal$$invokeCallback @ angular2-polyfills.js:1523lib$es6$promise$$internal$$publish @ angular2-polyfills.js:1494lib$es6$promise$$internal$$publishRejection @ angular2-polyfills.js:1444(anonymous function) @ angular2-polyfills.js:243run @ angular2-polyfills.js:138zoneBoundFn @ angular2-polyfills.js:111lib$es6$promise$asap$$flush @ angular2-polyfills.js:1305


And it still works fine if i'm building it to a single file with builder
** return builder.buildStatic('app/boot', 'build/boot.js') **
 Bundle contains my css(inlined) and everything is OK here (production).

-----------

Again second case - it works if i do it this way
**System.import("app/dashboard/dashboard.component.css!css");** (it didn't work with "./dashboard.component.css" syntax)
But works only in DEV, - this time i couldn't even do bundling - there are errors. Anyway, it would be better to always use **import "..."** syntax, without "System"

And below is simple config i used both in web and system-js-builder (i load it in gulp like this **var builder = new Builder('/','config.js');**)

System.config({
  baseURL: "/",
  defaultJSExtensions: true,
  transpiler: "typescript",
  paths: {
    "github:*": "jspm_packages/github/*",
    "npm:*": "jspm_packages/npm/*"
  },

  packages: {
    "app": {
      "format": "register",
      "defaultExtension": "js"
    }
  },
  map: {
    "angular2": "node_modules/angular2",
    "css": "github:systemjs/plugin-css@0.1.20",
    "rxjs": "node_modules/rxjs",
    "typescript": "npm:typescript@1.8.0"
  }
});


Thanks

**Comment by guybedford** [comment]
The first line of the error - `Error: http://sandbox.loc/app/dashboard/dashboard.component.css!http://sandbox.loc/jspm_packages/github/systemjs/plugin-css@0.1.20.js detected as register but didn't execute.(…)` still has the `(...)` included. Are you sure you expanded that before sharing the error message?

**Comment by mikhail-eremin** [comment]
Yes really, i missed some lines of error! :)
I've edited that message, now it's all i got in console

**Comment by guybedford** [comment]
Sorry, but it's a little bit tricky to follow all the directions here. Could you perhaps try one more time to share the simplest possible version of what is not working here? The more easily I can understand the exact steps and setup, the more easily I can assist.

**Comment by Emdot** [comment]
I have what _appears_ to be the same problem. My stack trace isn't from a minimized version, so maybe it's easier to work with. Here's a pretty minimal way to see it.

Stack:
```
"Error: http://localhost:51546/scripts/app/main.js detected as register but didn't execute.
   at Anonymous function (http://localhost:51546/scripts/systemjs/system.src.js:3287:11)
   at Anonymous function (http://localhost:51546/scripts/systemjs/system.src.js:3552:5)
   at Anonymous function (http://localhost:51546/scripts/systemjs/system.src.js:3800:7)
   at Anonymous function (http://localhost:51546/scripts/systemjs/system.src.js:4147:7)
   at Anonymous function (http://localhost:51546/scripts/systemjs/system.src.js:4383:9)
   at Anonymous function (http://localhost:51546/scripts/systemjs/system.src.js:346:9)
   at M (http://localhost:51546/scripts/systemjs/system-polyfills.js:4:8756)
   at H (http://localhost:51546/scripts/systemjs/system-polyfills.js:4:8343)
   at R.prototype.when (http://localhost:51546/scripts/systemjs/system-polyfills.js:4:12075)
   at b.prototype.run (http://localhost:51546/scripts/systemjs/system-polyfills.js:4:11109)"
```

Index.cshtml:
```
<!doctype html>
<html>
<head>
    <script src="/scripts/systemjs/system-polyfills.js"></script>
    <script src="/scripts/systemjs/system.src.js"></script>
    <script>
        System.config({
            baseURL: '/scripts',
            packages: {
                app: {
                    format: 'register',
                    defaultExtension: 'js'
                }
            }
        });
        System
            .import('app/main')
            .then(null, console.error.bind(console));
    </script>
</head>
<body>
    <h1>Test!</h1>
</body>
</html>
```

scripts/app/main.ts:
```
alert('yo');
````

The alert fires, and when I close it the error appears. If I remove the `format: 'register',` line the problem goes away. I don't actually know the implications of that line--I was following a tutorial for Angular 2.

**Comment by Emdot** [comment]
Aaand... now I can't reproduce it anymore, even with the test case I just gave. Make me wonder if there's some sort of browser-level caching problem.

Between when it broke and when it didn't I'd restored the full version, ran it, and deleted deleted some spurious js files that Visual Studio had built from my TypeScript but that weren't actually part of my build process.

**Comment by guybedford** [comment]
@Emdot right so the error is exactly correct since if it is trying to load `scripts/app/main.ts` as a System.register module, but the module is not a proper `System.register([], function(_export) { ... })` module form.

**Comment by akikoo** [comment]
Hi,

I have a related issue with CSS bundling in a Universal React app that I'm building. I want to import styles from components only in the browser (server import obviously doesn't work), but also bundle them for production (one bundle for JS and one for CSS). Later on, I'll be looking into splitting them to common and specific bundles as well, as described in the Builder documentation.

```js
// Using `import` in React component
// SystemJS Builder works both with `separateCSS` set to `true` and `false` 
// However I can't use this as components are initially rendered on the server
import 'dist/shared/components/Button/Button.css!';
```

```js
// Let's try to use `SystemJS.import` instead, importing styles only in the browser
// This works on DEV but SystemJS Builder completely ignores it
try {
  System.import('dist/shared/components/Button/Button.css!');
} catch (err) {}
```

Am I misunderstanding something here? Can this even be done currently? Should I approach this differently? How can I import styles in components (ignoring them on the server) and also bundle them for production?

Any pointers/thoughts would be highly appreciated.

Thanks!

**Comment by guybedford** [comment]
@akikoo you're right, CSS imports should be ignored when running on the server. I've created https://github.com/systemjs/plugin-css/issues/78 to track.

**Comment by akikoo** [comment]
Thanks @guybedford for the answer and for creating an issue. I'd imagine lots of people who are building universal apps are struggling with this. Can you give any estimate when this could be tackled?

Thanks again for all your amazing work!

**Comment by guybedford** [comment]
@akikoo this is all working fine actually. SystemJS builder doesn't currently trace dynamic imports like you have in the second code example although it should.

The way to do it is exactly to use the inline syntax `import 'dist/shared/components/Button/Button.css!';`. That will execute on the server fine without causing an error.

Then when you bundle for the client, the CSS will be bundled along with the rehydration code so everything works out fine.

**Comment by akikoo** [comment]
Hi again @guybedford, thanks for your assistance on this, I appreciate it! I've tried exactly the way you described above - also before writing here - but it's still not working for me. I'm happy to hear that it should work, I just can't figure out what I'm missing here. 

```js
// In the component
// Path is irrelevant here; if I import a JS module from that same directory it works fine
import 'dist/shared/components/Button/Button.css!';
// Other code
```
When running this on node/Express, an error is thrown with the following message: `Error: Cannot find module 'dist/shared/components/Button/Button.css!'`

So I'm only getting that error when importing a CSS file. Could this be a config/version issue? Any further ideas/help would be great, I'm pretty sure I'm close to getting this to work. 

Cheers

**Comment by guybedford** [comment]
@akikoo how are you running the code in Express? Note that for loading SystemJS modules in Node, you should be using `var System = require('systemjs'); System.config(...); System.import('app/main.js').then(...)` sort of style, as the Node module resolution is not compatible with SystemJS's module resolution.

**Comment by akikoo** [comment]
Hi @guybedford, 

I'm importing a Redux container in `server.js` and rendering that container to a string that is then sent to the client. On the client, in `main.js`, I'm rendering that same shared container. That container imports other, dumb React components where I want to import styles in each. 

On the server, I've tried importing those containers both with `import` and requiring `systemjs` as you suggested above. I made it work - for now - by importing all component styles in `main.js` instead. 

I'm afraid I have to park this now as I need to move on to other tasks. I've tried to find clear examples of the exact setup I need but somehow I can't make the pieces work together. It's probably because I seem to be lacking some fundamental understanding of universal module loading :) 

Thank you very much, again, for your time and for helping out!

I'll be sure to post here about the progress.
